Reorganize handleSyncOp

deitch opened the pull request:
Signed-off-by: Avi Deitcher <avi@deitcher.net>

This PR does several things to reorganize `pkg/pillar/cmd/downloader/`. The primary goal is to simplify the files and structure so that it will be easier to inject OCI registry downloading of artifacts for ECIs, while also standardizing how we download OCI images (currently for rkt, could be for runx or other runtimes).

This is _not_ complete, but a good start. Most of it is:

* separate a lot of `downloader.go`'s single 2000-line file into several files:
   * `syncOp.go` - the primary `handleSyncOp()` and `handleSyncOpResponse()`
   * `dns.go` - the DNS handlers
   * `rkt.go` - all of the rkt funcs, so that the only reference to it elsewhere is in `handleSyncOp()` with a single call to rkt fetch
* `status.go` - create an interface and implementation to pass to `handleSyncOp()`, so that it is ignorant of `status.Progress` or `publishDownloaderStatus()`, and instead just has an interface whose impl it can call `st.Progress(pct)` so it can update how much it is doing. This will make it easier to abstract out, for testing or using outside of pubsub for local downloads
* restructured a lot of `handleSyncOp()` to unify it, move some `switch` statements outside of the `for` loop, unified call, easier to understand.

The last two involve some logic change, so should be reviewed carefully. I don't think I missed anything, but that easily could mean I made some real mistakes.

I would like to break it up further, but this already helps; `downloader.go` is down in size by half, and I can see very clearly where to do both of the tasks I want to do:

* add pulling ECIs from an OCI registry
* restructure how `rkt` pulls images

kalyan-nidumolu commented on pkg/pillar/cmd/downloader/dns.go:
```diff
@@ -0,0 +1,42 @@
+		return
+	}
+	log.Infof("handleDNSModify for %s\n", key)
+	if cmp.Equal(ctx.deviceNetworkStatus, status) {
```
Do we need this check? PubSub already does this..

deitch replied:
Prob not, but this one I just moved into a separate file as is, since it was simple to do so. Can we leave it and clean up further next run?

kalyan-nidumolu replied:
Ok.. Is this just existing code moved - om ok with leaving it then..

kalyan-nidumolu commented on pkg/pillar/cmd/downloader/status.go:
```diff
@@ -0,0 +1,23 @@
+
+// DownloadStatus practical implementation of Status
+// requires a context and status
+type DownloadStatus struct {
```
Can we use a better name? We have the DownloaderContext as well as the DownloderStatus here.. Why do we need this struct? Can't we just pass these two separately where needed? This struct is a bit confusing..

deitch replied:
This is an example of where using interfaces can simplify our work. 

The previous download funcs, e.g. [doS3()](https://github.com/lf-edge/eve/blob/master/pkg/pillar/cmd/downloader/downloader.go#L1172), received `status *types.DownloaderStatus` as an arg, and then had to know the structure of it and to publish it. This made the download responsible for publishing, and being tightly tied to the structure.

By replacing that with an `interface` of `Status`, the download need only know it is receiving something that has a `Progress()` call. This completely decouples it. We can test `download()` without having it publish anything; we can connect it to an entirely different system of informing status updates.

The `DownloadStatus` is a practical implementation of `Status interface`, one that has a context and knows how to publish. 

Given that, I think a better name would be `PublishStatus`, since it is a practical implementation of `Status interface` that, well, publishes to `pubsub`. I will do that.

kalyan-nidumolu commented on pkg/pillar/cmd/downloader/syncop.go:
```diff
@@ -0,0 +1,272 @@
+	}
+
+	log.Debugf("handleSyncOp: config: %+v", config)
+	log.Debugf("handleSyncOp: IsContainer: %v", config.IsContainer)
```
We can remove this.. This will be printed as part of config anyways..

deitch replied:
Sure. Removing.

kalyan-nidumolu commented on pkg/pillar/cmd/downloader/download.go:
```diff
@@ -0,0 +1,87 @@
+				if osize != 0 {
+					percent := 100 * asize / osize
+					progress = uint(percent)
+				}
```
Can this be made into a utility routine for resp (zedUpload.DronaRequest ) ? 
   progress = resp.getProgress()

deitch replied:
I see no reason why not. 

This is another example of a place where an interface would be helpful, rather than a `struct`, allowing us to substitute other things that may simulate it.

I will get the `Progress()` in, and then we can see about interfacing that as well.

kalyan-nidumolu commented on pkg/pillar/cmd/downloader/download.go:
```diff
@@ -0,0 +1,87 @@
+				errStr := fmt.Sprintf("respChan EOF for <%s>, <%s>, <%s>",
+					dpath, region, filename)
+				log.Errorln(errStr)
+				return errors.New(errStr)
```
Hmm.. This means the download has failed.. right? We should propagate this error to zedcloud..

deitch replied:
I believe so, and we should propagate the error. Once again, though, the original code does this, see [here](https://github.com/lf-edge/eve/blob/master/pkg/pillar/cmd/downloader/downloader.go#L1148-L1151). Let's separate, then clean up.

kalyan-nidumolu commented on pkg/pillar/cmd/downloader/download.go:
```diff
@@ -0,0 +1,87 @@
+			}
+			if resp.IsError() {
+				return err
+			} else {
```
Since we are returning in the If block.. Else is notneeded? Can just use without indent?

deitch replied:
Correct. I am surprised linter didn't catch that (unless maybe we aren't using golangci-lint or similar?). Fixed.

kalyan-nidumolu commented on pkg/pillar/cmd/downloader/syncop.go:
```diff
@@ -0,0 +1,272 @@
+
+	if config.ImageSha256 != "" {
+		locFilename = locFilename + "/" + config.ImageSha256
+	}
```
What happens in the else case? Can we get a config with Empty SHA? What do we do in that case? Should we return an Error? Otherwise, we will just use the name locDirname/pending .. And if there are multiple such simultaneous downloads, we will run into errors..

Better return error check ( May be early in the process ) or just log.Fatal..

deitch replied:
Yes, that is what will happen, which is not good. However, this is another one of those, "this is how it worked before". Yes, it should be fixed, but let's iterate on it first. Separation here, fix those errors in several follow-ups. As long as we are not making it any worse, let's move ahead.

> Better return error check ( May be early in the process ) or just log.Fatal

Part of the issue is that `handleSyncOp()` doesn't return _anything_, instead it either does `log.Fatalf()`, or it calls `handleSyncOpResponse()` (or both). Using `log.Fatalf()` is not a good move here (best handled only in main functions, or places that are meant to work with them), but, again, one step at a time.

deitch commented:
I addressed the issues you raised. Some punted on just so we can get the restructure in place quickly at the current level (as a minimum), some addressed directly, see the comments.

The other issue you raised is:

> can add some go unit tests for some of these Smaller routines

No objection, but again, let's separate concerns. First break it out as is, then improve it in other PRs. The other reason to defer is that the funcs are somewhat tightly interwoven; that `status`/`Progress()` one is a good example. We need a few more of those to make testing easier.

deitch commented:
Also, we need to do more teasing out of these; `downloader.go` remains 1100+ lines. But it is a start.

deitch commented:
@rvs or @nordmark: Kalyan is good with it. Let’s get it merged in.

deitch commented:
@kalyan-nidumolu: can you open two issues around those two points you raised above, so it doesn’t slip through in the future?

deitch commented:
How odd. yetus failed with the odd message:

> Root in master failed
>   Artifact not found

deitch commented:
Urgh. That keeps happening. I’ll update it (again), and we need to get it merged in as soon as possible after that so it doesn’t happen again.
